scripts/publish.rs

```diff
@@ -0,0 +1,318 @@
+                    .trim()
+                    .to_string(),
+            );
+        }
+        if version.is_none() && line.starts_with("version = \"") {
+            version = Some(
+                line.replace("version = \"", "")
+                    .replace("\"", "")
+                    .trim()
+                    .to_string(),
+            );
+        }
+        if line.starts_with("publish = false") {
+            publish = false;
+        }
+    }
+    let name = name.unwrap();
+    let version = version.unwrap();
+    let next_version = if CRATES_TO_PUBLISH.contains(&&name[..]) {
+        bump(&version)
+    } else {
+        version.clone()
+    };
+    if name == "witx" {
+        publish = false;
+    }
+    Crate {
+        manifest: manifest.to_path_buf(),
+        name,
+        version,
+        next_version,
+        publish,
+    }
+}
+
+fn bump_version(krate: &Crate, crates: &[Crate]) {
+    let contents = fs::read_to_string(&krate.manifest).unwrap();
+
+    let mut new_manifest = String::new();
+    let mut is_deps = false;
+    for line in contents.lines() {
+        let mut rewritten = false;
+        if !is_deps && line.starts_with("version =") {
+            if CRATES_TO_PUBLISH.contains(&&krate.name[..]) {
+                println!(
+                    "bump `{}` {} => {}",
+                    krate.name, krate.version, krate.next_version
+                );
+                new_manifest.push_str(&line.replace(&krate.version, &krate.next_version));
+                rewritten = true;
+            }
+        }
+
+        is_deps = if line.starts_with("[") {
+            line.contains("dependencies")
+        } else {
+            is_deps
+        };
+
+        for other in crates {
+            if !is_deps || !line.starts_with(&format!("{} ", other.name)) {
+                continue;
+            }
+            if !line.contains(&other.version) {
+                if !line.contains("version =") {
+                    continue;
+                }
+                panic!(
+                    "{:?} has a dep on {} but doesn't list version {}",
+                    krate.manifest, other.name, other.version
+                );
+            }
+            rewritten = true;
+            new_manifest.push_str(&line.replace(&other.version, &other.next_version));
+            break;
+        }
+        if !rewritten {
+            new_manifest.push_str(line);
+        }
+        new_manifest.push_str("\n");
+    }
+    fs::write(&krate.manifest, new_manifest).unwrap();
+}
+
+fn bump(version: &str) -> String {
+    let mut iter = version.split('.').map(|s| s.parse::<u32>().unwrap());
+    let major = iter.next().expect("major version");
+    let minor = iter.next().expect("minor version");
+    let patch = iter.next().expect("patch version");
+    if major != 0 {
+        format!("{}.0.0", major + 1)
+    } else if minor != 0 {
+        format!("0.{}.0", minor + 1)
+    } else {
+        format!("0.0.{}", patch + 1)
+    }
```

Comment:
I may be missing something here, but I don't see what this function is supposed to be doing.

Reply:
I added some more documentation to this, but the intention is to basically figure out the next version number we're going to write down everywhere for each crate individually.

Reply:
Ah, makes sense now. Thanks!